include/souper/Infer/InstSynthesis.h

```diff
@@ -0,0 +1,300 @@
+  Inst::Phi,
+  // Disabled due to scalability challenges
+  Inst::AddNSW, Inst::AddNUW, Inst::AddNW,
+  Inst::SubNSW, Inst::SubNUW, Inst::SubNW,
+  Inst::MulNSW, Inst::MulNUW, Inst::MulNW,
+  Inst::ShlNSW, Inst::ShlNUW, Inst::ShlNW,
+};
+
+/// Supported component library.
+/// The width of most of the instructions is unknown in advance, therefore, we
+/// indicate this by setting the Width to 0. Per each input of distinct width,
+/// a component of that width is instantiated.
+/// Again, note that constants are treated as ordinary inputs
+static const std::vector<Component> CompLibrary = {
+  Component{Inst::Add, 0, {0,0}},
+  Component{Inst::Sub, 0, {0,0}},
+  Component{Inst::Mul, 0, {0,0}},
+  Component{Inst::UDiv, 0, {0,0}},
+  Component{Inst::SDiv, 0, {0,0}},
+  Component{Inst::UDivExact, 0, {0,0}},
+  Component{Inst::SDivExact, 0, {0,0}},
+  Component{Inst::URem, 0, {0,0}},
+  Component{Inst::SRem, 0, {0,0}},
+  Component{Inst::And, 0, {0,0}},
+  Component{Inst::Or, 0, {0,0}},
+  Component{Inst::Xor, 0, {0,0}},
+  Component{Inst::Shl, 0, {0,0}},
+  Component{Inst::LShr, 0, {0,0}},
+  Component{Inst::LShrExact, 0, {0,0}},
+  Component{Inst::AShr, 0, {0,0}},
+  Component{Inst::AShrExact, 0, {0,0}},
+  Component{Inst::Select, 0, {1,0,0}},
+  Component{Inst::Eq, 1, {0,0}},
+  Component{Inst::Ne, 1, {0,0}},
+  Component{Inst::Ult, 1, {0,0}},
+  Component{Inst::Slt, 1, {0,0}},
+  Component{Inst::Ule, 1, {0,0}},
+  Component{Inst::Sle, 1, {0,0}},
+  //
+  Component{Inst::CtPop, 0, {0}},
+  Component{Inst::BSwap, 0, {0}},
+  Component{Inst::Cttz, 0, {0}},
+  Component{Inst::Ctlz, 0, {0}}
+};
+
+class InstSynthesis {
+public:
+  // Synthesize an instruction from the specification in LHS
+  std::error_code synthesize(SMTLIBSolver *SMTSolver,
+                             const BlockPCs &BPCs,
+                             const std::vector<InstMapping> &PCs,
+                             Inst *TargetLHS, Inst *&RHS,
+                             InstContext &IC, unsigned Timeout);
+
+private:
+  /// LHS to synthesize
+  Inst *LHS;
+  /// Replacement context for inst printing
+  ReplacementContext Context;
+  /// Components to be used
+  std::vector<Component> Comps;
+  /// Constant components
+  std::vector<Component> ConstComps;
+  /// Program inputs
+  std::vector<Inst *> Inputs;
+  std::set<unsigned> InputWidths;
+  /// Input location set I
+  std::vector<LocInst> I;
+  /// Component input location set P
+  std::vector<LocInst> P;
+  /// Component output location set R
+  std::vector<LocInst> R;
+  /// Output location O
+  LocInst O;
+  /// Location variables set L (I \cup P \cup R \cup {O}).
+  std::vector<LocInst> L;
+  /// Number of input variables
+  unsigned N;
+  /// Number of components + N
+  unsigned M;
+  /// A mapping from a location variable to a concrete component instance,
+  /// namely created instruction
+  std::map<LocVar, Inst *> CompInstMap;
+  /// Location variable's width is fixed
+  const unsigned LocInstWidth = 32;
```

Comment:
Is there any reason to use so many bits here? Might it be helpful to reduce this to 6 or 8?